src/linux_mcp_server/tools/system_info.py

```diff
@@ -350,156 +511,272 @@ async def get_memory_information(
+                # Handle both success and failure (file may not exist)
+                value = stdout.strip() if returncode == 0 and stdout else ""
+                results.append(f"{attr}:{value}")
+
+            # Combine into single cache entry
+            combined_output = "\n".join(results)
+            cache[("usb_attrs", device_name)] = combined_output
+
+    return cache
+
+
+def _parse_pci_device_attrs(address: str, output: str) -> PCIDevice:
+    """Parse PCI device attributes from raw output."""
+    device = PCIDevice(address=address)
+
+    for line in output.strip().split("\n"):
+        if ":" in line:
+            key, value = line.split(":", 1)
+            value = value.strip()
+            if value:
+                match key:
+                    case "vendor":
+                        device.vendor_id = value
+                    case "device":
+                        device.device_id = value
+                    case "class":
+                        device.class_id = value
+                    case "label":
+                        device.description = value
+
+    return device
+
+
+def _parse_usb_device_attrs(device_name: str, output: str) -> USBDevice:
+    """Parse USB device attributes from raw output."""
+    parts = device_name.split("-")
+    bus = parts[0]
+    device = USBDevice(bus=bus, device=device_name)
+
+    for line in output.strip().split("\n"):
+        if ":" in line:
+            key, value = line.split(":", 1)
+            value = value.strip()
+            if value:
+                match key:
+                    case "idVendor":
+                        device.vendor_id = value
+                    case "idProduct":
+                        device.product_id = value
+                    case "product":
+                        device.description = value
+                    case "manufacturer":
+                        device.manufacturer = value
+
+    return device
+
+
+async def _parse_device_information(raw_outputs: dict[CommandKey, RawCommandOutput]) -> ParsedData:
+    """
+    Parse phase: Convert raw device outputs into structured device data.
+
+    Returns a dictionary with pci_devices and usb_devices lists.
+    """
+    pci_devices = []
+    usb_devices = []
+
+    # Parse PCI and USB devices
+    for command_key, output in raw_outputs.items():
+        if command_key[0] == "pci_attrs" and len(command_key) == 2:
+            address = command_key[1]
+            pci_devices.append(_parse_pci_device_attrs(address, output))
+        elif command_key[0] == "usb_attrs" and len(command_key) == 2:
+            device_name = command_key[1]
+            usb_devices.append(_parse_usb_device_attrs(device_name, output))
+
+    return {
+        "pci_devices": pci_devices,
+        "usb_devices": usb_devices,
+    }
+
+
+async def _filter_device_information(
+    parsed_data: ParsedData, device_types: list[str] | None = None, limit: int | None = None
+) -> ParsedData:
```

Comment:
For all these filter functions, we could use the [validation decorator](https://docs.pydantic.dev/latest/concepts/validation_decorator/#using-the-field-function-to-describe-function-parameters) to more strictly enforce the function inputs and outputs. This would allow us to use mutable defaults, something that's not possible to do safely by default.

```suggestion
DeviceTypes(StrEnum):
    pci = "pci"
    usb = "usb"
    
@validate_call
async def _filter_device_information(
    parsed_data: ParsedData,
    device_types: list[DeviceTypes] | None = None, 
    limit: t.Annotated[int | None, Field(lt=1_0000)] = None,
) -> ParsedData:
```